app/app.go

```diff
@@ -519,28 +523,37 @@ func (app *App) BeginBlocker(ctx sdk.Context, req abci.RequestBeginBlock) abci.R
 func (app *App) EndBlocker(ctx sdk.Context, req abci.RequestEndBlock) abci.ResponseEndBlock {
 	res := app.mm.EndBlock(ctx, req)
 	// NOTE: this is a specific feature for upgrading from v1 to v2. It will be deprecated in v3
-	if app.UpgradeKeeper.ShouldUpgradeToV2(req.Height) {
-		if app.AppVersion(ctx) == v1.Version {
-			app.SetAppVersion(ctx, v2.Version)
+	if app.UpgradeKeeper.ShouldUpgradeToV2(req.Height) && app.AppVersion(ctx) == v1.Version {
+		if err := app.Upgrade(ctx, v2.Version); err != nil {
+			panic(err)
 		}
 		// from v2 to v3 and onwards we use a signalling mechanism
 	} else if shouldUpgrade, version := app.UpgradeKeeper.ShouldUpgrade(); shouldUpgrade {
 		// Version changes must be increasing. Downgrades are not permitted
 		if version > app.AppVersion(ctx) {
-			app.SetAppVersion(ctx, version)
-			app.UpgradeKeeper.ResetTally(ctx, version)
+			if err := app.Upgrade(ctx, version); err != nil {
+				panic(err)
+			}
 		}
 	}
 	return res
 }
 
+func (app *App) Upgrade(ctx sdk.Context, version uint64) error {
+	app.SetAppVersion(ctx, version)
+	return app.mm.RunMigrations(ctx, app.configurator, version, version)
```

Comment:
[question] why is `version` used for both `fromVersion` and `toVersion`? If we're upgrading, from a version to a different version, I would expect these two versions to be distinct.

Can a clarifying comment be added?

Reply:
Changed it